File: bill_buddy_backend/src/model/bill.js
const mongoose = require('mongoose');

const billSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },

    userBillerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'UserBiller',
      required: true,
    },

    billerName: {
      type: String,
      required: true,
      trim: true,
    },

    nickname: {
      type: String,
      required: true,
      trim: true,
    },

    category: {
      type: String,
      required: true,
    },

    accountDetails: {
      type: Map,
      of: String,
      required: true,
    },

    amount: {
      type: Number,
      required: true,
      min: 0,
    },

    dueDate: {
      type: Date,
      required: true,
    },

    billingPeriod: {
      type: String,
      required: true,
    },

    status: {
      type: String,
      enum: [
        'upcoming',
        'due_soon',
        'overdue',
        'paid',
      ],
      default: 'upcoming',
    },
  },
  {
    timestamps: true,
  },
);

module.exports = mongoose.model(
  'Bill',
  billSchema,
  'bills',
);

File: bill_buddy/lib/models/bill.dart
class Bill {
  final String id;
  final String userBillerId;
  final String billerName;
  final String nickname;
  final String category;
  final Map<String, dynamic> accountDetails;
  final double amount;
  final DateTime dueDate;
  final String billingPeriod;
  final String status;

  const Bill({
    required this.id,
    required this.userBillerId,
    required this.billerName,
    required this.nickname,
    required this.category,
    required this.accountDetails,
    required this.amount,
    required this.dueDate,
    required this.billingPeriod,
    required this.status,
  });

  factory Bill.fromJson(Map<String, dynamic> json) {
    return Bill(
      id: json['_id'] as String,
      userBillerId: json['userBillerId'] as String,
      billerName: json['billerName'] as String,
      nickname: json['nickname'] as String,
      category: json['category'] as String,
      accountDetails: Map<String, dynamic>.from(
        json['accountDetails'] ?? {},
      ),
      amount: (json['amount'] as num).toDouble(),
      dueDate: DateTime.parse(json['dueDate'] as String),
      billingPeriod: json['billingPeriod'] as String,
      status: json['status'] as String,
    );
  }
}